refactor: nextcord.Interaction subclassing

Gangsta-Muffin opened the pull request:
## Summary
PR for #952 

Create Subclasses for `nextcord.Interaction`

Subclasses will include:
-------------------------

`ApplicationCommandInteraction`
Interaction object for application commands

`ApplicationAutocompleteInteraction`
Interaction object for slash command autocomplete

`ViewInteraction`
Interaction object for message components

`ModalSubmitInteraction`
Interaction object for Modal submits

Following object names will/may still change:
- `ApplicationAutocompleteInteraction` 
- `ViewInteraction` to `MessageComponentInteraction`


## Checklist (currently incomplete)

- [x] SlashApplicationInteraction (Interaction subclass)
- [x] ApplicationAutocompleteInteraction (Interaction subclass)
- [x] ViewInteraction (Interaction subclass)
- [x] ModalSubmitInteraction (Interaction subclass)
- [x] InteractionResponse subclasses
- [x] Interaction subclasses in `nextcord.types.interactions`
-------------

- [x] Update external Modules to use/identify correct Object
- [x] Add new subclasses to `__all__` Set in interactions.py
- [x] Update docs
- [x] Update additional stuff (for example: `__slots__`, type hints, bugs, ...)

-------------
- [ ] Final test (for bugs) - fix if necessary

Gangsta-Muffin committed:
class creation

Creat first classes + Create PR

Gangsta-Muffin committed:
Merge pull request #1 from Gangsta-Muffin/interaction_subclasses

class creation

Gangsta-Muffin committed:
First Implementation of interaction_subclasses

* Check and Use Interaction type in Client.py
* Import Interaction subclasses in Client.py
* Transfer SlashApplicationInteraction specific data from Base to subclass
* Copy SlashApplicationInteraction contents to ApplicationAutocompleteInteraction (use same data)

Gangsta-Muffin committed:
First implementation of Interaction subclasses

* Check and use correct Interaction type in Client.py
* Add imports for Interaction subclasses in Client.py

* Transfer SlashApplicationInteraction specific data from Interaction (parent class)
* Transfer ApplicationAutocompleteInteraction specific data from Interaction (parent class)
* Add subclass refined properties to SlashApplicationInteraction
* Add subclass refined properties to ApplicationAutocompleteInteraction
* Create PartialInteractionOption (will contain options data for SlashApplicationInteraction and ApplicationAutocompleteInteraction)

Gangsta-Muffin committed:
First implementation of Interaction subclasses

* Check and use correct Interaction type in Client.py
* Add imports for new Interaction subclasses

* Transfer SlashApplicationInteraction specific data from Interaction (parent class)
* Transfer ApplicationAutocompleteInteraction specific data from Interaction (parent class)
* Add properties for SlashApplicationInteraction refined data
* Add properties for ApplicationAutocompleteInteraction refined data

Gangsta-Muffin committed:
Merge branch 'master' of https://github.com/Gangsta-Muffin/interacion_subclasses

pre-commit-ci[bot] committed:
style: auto fixes from pre-commit hooks

Gangsta-Muffin committed:
feat: Create SlashOptionData and Add support for custom Interaction subclasses

* Add parameters to get_interaction to support custom Interaction subclasses in Client.py
* Create SlashOptionData to hold recieved SlashOptionData - to be used in Interaction classes (Interactions.py)

* Rename SlashApplicationCommand subinteraction class to SlashApplicationInteraction
* Add option data handling in SlashApplicationInteraction (using SlashOptionData) (new .options attribute)

* Add option data handlin in ApplicationAutocompleteInteraction (using SlashOptionData) (new .options attribute)
* Add ocused_option attribute to ApplicationAutocompleteInteraction (fast way to get Slashoption invoking Autocomplete)

Gangsta-Muffin committed:
perf: Stop double initialization of focused options

Stop double Initialization for focused slash options - instead append the first initialization to collection list

Gangsta-Muffin committed:
feat: Create SlashOptionData and Add support for custom Interaction subclasses

* Add parameters to get_interaction to support custom Interaction subclasses in Client.py
* Create SlashOptionData to hold recieved SlashOptionData - to be used in Interaction classes (Interactions.py)

* Rename SlashApplicationCommand subinteraction class to SlashApplicationInteraction
* Add option data handling in SlashApplicationInteraction (using SlashOptionData) (new .options attribute)

* Add option data handlin in ApplicationAutocompleteInteraction (using SlashOptionData) (new .options attribute)
* Add ocused_option attribute to ApplicationAutocompleteInteraction (fast way to get Slashoption invoking Autocomplete)

pre-commit-ci[bot] committed:
style: auto fixes from pre-commit hooks

Gangsta-Muffin committed:
refactor: Transfer View and Modal specific data and methods to their subclasses

* Copy Interaction.edit() method to ViewInteraction
* Copy Intenteraction.edit() to ModalSubmitInteraction
* Remove Interaction.edit() from Interaction

* Remove .message attribute from Interaction
* Add .message attribute to ViewInteraction
* Add .message attribute to ModalSubmitInteraction

* Add attributes for ModalSubmit refined data (ModalSubmitInteraction)
* add attributes for MessageComponent refined data (ViewInteraction)

* Add comments for questioning changes

pre-commit-ci[bot] committed:
style: auto fixes from pre-commit hooks

Gangsta-Muffin committed:
fix: interaction.edit() method fix

* Fixed ViewInteraction.edit() method - can now also edit ephemeral messages
* Fixed ModalSubmitInteraction.edit() method - can now also edit ephemeral messages

Refactor:
* Remove  prints from testing
* Add comments for future changes
* Add __repr__ method to SlashOptionData class

pre-commit-ci[bot] committed:
style: auto fixes from pre-commit hooks

Gangsta-Muffin committed:
refactor: Add classes to __all__

* Add classes to __all__ (interactions.py)
* removed prints from testing

Gangsta-Muffin committed:
docs: typing: refactor: Class descriptions, correct typing and attrs updates

* Change SlashApplicationInteraction name to ApplicationCommandInteraction
* Type the correct Interaction subclass in client.py

* Add class Descriptions to Interaction Subclasses
* Correct __slots__ in Interaction class
* Ad __slots__ to Interaction subclasses
* Change some attribute names to have more appropriate name

pre-commit-ci[bot] committed:
style: auto fixes from pre-commit hooks

EmreTech commented on nextcord/interactions.py:
```diff
@@ -559,6 +539,485 @@ async def send(
+        self.app_command_name: str = self.data["name"]  # type: ignore # self.data should be present here
+        self.app_command_id: int = self.data["id"]  # type: ignore # self.data should be present here
+
+        try:
+            self.options: list = self._get_application_options(self.data["options"])  # type: ignore # self.data should be present here
+        except KeyError:
+            self.options: list = []
```
Fully define this type hint.

EmreTech commented on nextcord/interactions.py:
```diff
@@ -559,6 +539,485 @@ async def send(
+    application_command: Union[SlashApplicationSubcommand, BaseApplicationCommand]
+        The application command that triggered the interaction
+    app_command_name: :class:`str`
+        The name of the application command that triggered the interaction
```
```suggestion
        The name of the application command that triggered the interaction.
```

EmreTech commented on nextcord/interactions.py:
```diff
@@ -559,6 +539,485 @@ async def send(
+    token: :class:`str`
+        The token to continue the interaction. These are valid
+        for 15 minutes.
+    data: :class:`dict`
```
It would be nice if this `data` attribute was uniquely typed for each interaction subclass to use a corresponding typed dict.

ooliver1 replied:
Issue is, `nextcord.types` in general isn't documented, would it make more sense to simply link to the discord docs?

EmreTech replied:
I suppose.

Skelmis replied:
As an side, possibly we should discuss introducing something like `discord-typings` if we want to go that route? @ooliver1

ooliver1 replied:
`discord-typings` too does not have documentation

EmreTech commented on nextcord/interactions.py:
```diff
@@ -559,6 +539,485 @@ async def send(
 
+
+class ApplicationCommandInteraction(Interaction):
+    """Represents the base class for all Discord interactions.
```
```suggestion
    """Represents an application command interaction.
```

EmreTech commented on nextcord/interactions.py:
```diff
@@ -559,6 +539,485 @@ async def send(
+
+
+class ApplicationAutocompleteInteraction(Interaction):
+    """Represents the base class for all Discord interactions.
```
```suggestion
    """Represents an autocomplete application command interaction.
```

EmreTech commented on nextcord/interactions.py:
```diff
@@ -559,6 +539,485 @@ async def send(
+        return options_collection
+
+
+class ApplicationAutocompleteInteraction(Interaction):
```
If this class is going to copy a lot of logic from `ApplicationCommandInteraction`, shouldn't this class be a subclass of `ApplicationCommandInteraction`?

EmreTech commented on nextcord/interactions.py:
```diff
@@ -559,6 +539,485 @@ async def send(
+    attached: :class:`InteractionAttached`
+        The attached data of the interaction. This is used to store any data you may need inside the interaction for convenience. This data will stay on the interaction, even after a :meth:`Interaction.application_command_before_invoke`.
+    message: :class:`Message`
+        The message the component is attached to
```
```suggestion
        The message the component is attached to.
```

EmreTech commented on nextcord/interactions.py:
```diff
@@ -559,6 +539,485 @@ async def send(
+    message: :class:`Message`
+        The message the component is attached to
+    component_id: :class:`int`
+        The ID of the component that triggered the Interaction
```
```suggestion
        The ID of the component that triggered the interaction.
```

EmreTech commented on nextcord/interactions.py:
```diff
@@ -559,6 +539,485 @@ async def send(
+    component_id: :class:`int`
+        The ID of the component that triggered the Interaction
+    value: :class:`str`
+        The value of the component that trgigered the interaction
```
```suggestion
        The value of the component that triggered the interaction.
```

EmreTech commented on nextcord/interactions.py:
```diff
@@ -559,6 +539,485 @@ async def send(
+        "_cs_channel",
+    )
+
+    def __init__(self, *, data: InteractionPayload, state: ConnectionState):
```
It would also be nice if the `data` parameter was more specifically typed. Right now, it allows for any interaction payload, but in this case we want to limit it to application command interaction payloads.

EmreTech commented on nextcord/interactions.py:
```diff
@@ -559,6 +539,485 @@ async def send(
+            Union[SlashApplicationSubcommand, BaseApplicationCommand]
+        ] = None
+
+    def _from_data(self, data: InteractionPayload):
```
Same here.

EmreTech commented on nextcord/interactions.py:
```diff
@@ -559,6 +539,485 @@ async def send(
+        The attached data of the interaction. This is used to store any data you may need inside the interaction for convenience. This data will stay on the interaction, even after a :meth:`Interaction.application_command_before_invoke`.
+    message: :class:`Message`
+        The message the component is attached to
+    component_id: :class:`int`
```
It looks like the component id attribute is really the custom id. Custom ids for components are typed to be strings, not integers.

EmmmaTech reviewed (changes requested):
Two more things that I decided to leave out of individual review comments:
- The rest of the library isn't typed to use these new subclasses.
- It would be a lot cleaner imo if the interaction subclasses were split into multiple files under an `interactions` folder. This also includes the base interaction class.

FYI, some of the review comments I specifically made on `ApplicationCommandInteraction` apply to the rest of the subclasses, I didn't want to copy all of those review comments over and over again.

MaskDuck reviewed (changes requested):
also i'm unsure whether should we keep documenting Interaction -- as it doesn't have any real-world usage now.

i think it's a bit breaking but fine, acceptable

Gangsta-Muffin committed:
Refactor: update changes mentioned in comments

* Remove parent attributes from Interaction subclass descriptions (suggestion from PR comments)
* Remove 'version added' in Interaction subclasses (suggestion from PR comments)
* Correct spelling and punctuation in Interaction subclass descriptions (suggestion from PR comments)
* ApplicationAutocompleteInteraction now inherits from ApplicationCommandInteraction
* Correct list typehinting for application options list in ApplicationCommandInteraction

pre-commit-ci[bot] committed:
style: auto fixes from pre-commit hooks

MaskDuck reviewed (changes requested):
would need to add backward compatibilities or fix pyright error
i actually prefer add backward compatibilities bcuz it would not take much time

but you also need to re-typehint the whole lib where just use interactions, thx =))

Gangsta-Muffin committed:
typing: update interaction type hints; refactor: new dedicated interactions folder

Gangsta-Muffin committed:
typing: update interaction type hints; refactor: new dedicated interactions folder

Gangsta-Muffin committed:
Merge branch 'master' of https://github.com/Gangsta-Muffin/interacion_subclasses

Gangsta-Muffin committed:
refactor: typing: Interaction Class seperation

* Create dedicated folder for all Interaction classes and subclasses
* Split interaction subclasses into dedicated files
* Change import locations and imported items to import from new interaction folder and dedicated interactions file (for interaction subclass)
* Change type annotations to dedicated subclasses (unfinished)

pre-commit-ci[bot] committed:
style: auto fixes from pre-commit hooks

Gangsta-Muffin committed:
docs: typing: refactor: Write last class description, add typing, work on circular import error

* Write class description for SlashOptionData
* Add type hints in SlashOptionData
* Work on circular import error: Edit imports

pre-commit-ci[bot] committed:
style: auto fixes from pre-commit hooks

EmmmaTech reviewed (commented):
I would rename the submodules of `nextcord.interactions` to remove the `_interaction(s)` suffix because the name of the parent module already implies that the submodules contain interaction subclasses.

A good example of where this was done right is `nextcord.ui.select`.

Gangsta-Muffin committed:
refactor: update interaction submodule names

* Update file names for Interaction subclasses -> suggestion by EmreTech

Skelmis committed:
fix: circular imports

Skelmis commented on nextcord/interactions.py:
```diff
@@ -594,7 +799,120 @@ async def edit(self, *args, **kwargs) -> Optional[Message]:
+
+        .. describe:: x != y
+
+            Checks if two interactions are not equal.
```
No, the current documented addition is consistent with the rest of the library

Skelmis commented on nextcord/interactions.py:
```diff
@@ -594,7 +799,120 @@ async def edit(self, *args, **kwargs) -> Optional[Message]:
+
+        .. describe:: x == y
+
+            Checks if two interactions are equal.
```
No, the current documented addition is consistent with the rest of the library

Skelmis commented on nextcord/interactions.py:
```diff
@@ -1301,3 +1619,23 @@ class InteractionMessage(_InteractionMessageMixin, Message):
+    """Class description"""
+
+    def __init__(self, data) -> None:
+        # Still missing `options?` Data from (view Discord Docs)
```
What do you mean "add the property", I am unsure on what is required to resolve this review comment

Gangsta-Muffin committed:
fix: Fix last import errors

* Fix remaining import errors caused by attempts from trying to fix circular import error

Gangsta-Muffin committed:
refactor: Change class name and fix __all__ set

* Change ViewInteraction to MessageComponentInteraction
* update __all__ set in application interaction file to include SlashOptionData

Gangsta-Muffin committed:
fix: Fix subcommand interaction.options

* Fix the options attribute for sub-/groupcommands in ApplicationCommandInteraction.
* Add __slots__ where they were missing

Gangsta-Muffin committed:
Docs: Fix: Update docs for Client.get_interaction and Fix bugs and other

* Update documentation for Client.get_interaction
* Fix bugs in ApplicationCommandInteraction._get_application_options - Now returns empty list if no options were provided
* Fix Support for custom interaction classes - Warn user if they set a custom interaction class to the 'cls' parameter: Parameter is only used when the interaction type wasn't recognized

Gangsta-Muffin committed:
Docs: resolve conversation

* Update data attribute description to include related discord developer portal link in SlashOptionData class
* Update data attribute description to include related descord developer portal link in Interaction class

Gangsta-Muffin committed:
Docs: update documentation for Client.get_interaction()

* Add code example
* Add documentation for warnings
* Elaborate on the description for the cls parameter

Gangsta-Muffin committed:
Merge: Merg latest nextcord version into PR

pre-commit-ci[bot] committed:
style: auto fixes from pre-commit hooks

EmreTech commented on nextcord/interactions/message_component.py:
```diff
@@ -0,0 +1,109 @@
+class MessageComponentInteraction(Interaction):
+    """Represents the interaction for all messsage components.
+
+    This interaction is get's triggered by :class:`nextcord.ui.View`
```
```suggestion
    This interaction gets triggered by a :class:`nextcord.ui.View`.
```

EmreTech commented on nextcord/interactions/message_component.py:
```diff
@@ -0,0 +1,109 @@
+    def _from_data(self, data: InteractionPayload) -> None:
+        super()._from_data(data=data)
+
+        message = data["message"]  # type: ignore # should be present here - not sure why its causing issues
```
According to the type hint for `data` in this function (which is `nextcord.types.Interaction`), the `message` key might not be available because `message` in the `Interaction` typed dict is not required.

EmreTech commented on nextcord/interactions/message_component.py:
```diff
@@ -0,0 +1,109 @@
+        ) or Message(  # ---- check for and create correct message object ----
+            state=self._state, channel=self.channel, data=message  # type: ignore
+        )
+        self.component_id = self.data["custom_id"]  # type: ignore # self.data should be present here
```
`self.data` is typed as optional, so according to Pyright it could not be present.

EmreTech commented on nextcord/state.py:
```diff
@@ -1377,11 +1377,13 @@ def parse_interaction_create(self, data) -> None:
             component_type = interaction.data["component_type"]  # type: ignore
-            self._view_store.dispatch(component_type, custom_id, interaction)
+            self._view_store.dispatch(component_type, custom_id, interaction)  # type: ignore
+            # Some Interaction types incompatible with this method -> accounted for via if-statement
```
Not in Pyright's eyes. Checking if a certain value in the raw interaction data used to make an `Interaction` is another value doesn't mean that the created object is magically that type. You have to add more to the if statement to justify this.

EmreTech commented on nextcord/ui/item.py:
```diff
@@ -4,7 +4,7 @@
 from typing import TYPE_CHECKING, Any, Callable, Coroutine, Generic, Optional, Tuple, TypeVar
 
-from ..interactions import ClientT, Interaction
+from ..interactions.base import ClientT, Interaction
```
Shouldn't this be using `MessageComponentInteraction`?

EmreTech commented on nextcord/interactions/modal_submit.py:
```diff
@@ -0,0 +1,106 @@
+class ModalSubmitInteraction(Interaction):
+    """Represents the interaction for all modal submits.
+
+    This interaction gets triggered by :class:`nextcord.ui.Modal`
```
```suggestion
    This interaction is triggered by a :class:`nextcord.ui.Modal`.
```

EmreTech commented on nextcord/interactions/application.py:
```diff
@@ -0,0 +1,177 @@
+
+
+# class is here due to circular import issues
+class SlashOptionData:
```
Why is there documentation for this class? If it's copy-pasted here for circular import issues, then it would be internally used.

Gangsta-Muffin replied:
The point behind this comment was the fact that I was originally trying to add the class to `nextcord.application_command` since there is a class of similar nature in that file: `SlashOption` (line 1218 ff.). The `SlashOptionData` class is designed to contain SlashOption data provided by a user, while the `SlashOption` class is designed to take requirements for an application command option - requirements set by the developer.

Gangsta-Muffin replied:
conversation continueing in other conversation

EmreTech commented on nextcord/interactions/application.py:
```diff
@@ -0,0 +1,177 @@
+        options_collection: List[SlashOptionData] = []
+
+        for option in options:
+            options_collection.append(SlashOptionData(option))
```
There's an easy way to get around the circular import issues if this is what `SlashOptionData` is being used for in the code. Just import whatever you need in the scope of the function. 
For example:

```python
def foo():
    from .bar import soap
    soap()
```

And for any time `SlashOptionData` is being used for type hints outside the scope of a function, import it under a `TYPE_CHECKING` block and enable `from __future__ import annotations`.

Gangsta-Muffin replied:
I assume `foo()` represents the `_get_application_options()` method in this case? It is customary to have all imports at the top of a python file. Is it really a good idea to import within a function mid file? 

(Yes, this is why the class is in this file- was originally going to put it in `nextcord.application_command.py`. Since you asked about this in the other comment, and the discussion will continue here, I shall resolve that conversation)

EmmmaTech reviewed (changes requested):
Additionally, every new submodule in the `nextcord.interactions` module is missing the license identifier comment (`# SPDX-License-Identifier: MIT`).

Gangsta-Muffin committed:
Docs: typo correction

Co-authored-by: Emre Terzioglu <50607143+EmreTech@users.noreply.github.com>

Gangsta-Muffin committed:
Fixed Issues found in EmreTech's review

- Added interaction Subclasses in types.interactions
- Retyped all Interaction subclasses to use correct types.interaction class
- Moved SlashOptionData to nextcord.application_command
- Added SlashOptionData import in method to avoid circular import issues
- Check interaction type for view dispatch in nextcord.state instead of checking api data (partially due to pyright)
- added # SPDX-License-Identifier: MIT license identifier commend to the top of submodules in nextcord.interactions

Gangsta-Muffin committed:
Resolving Conflicts to Nextcord

pre-commit-ci[bot] committed:
style: auto fixes from pre-commit hooks

Gangsta-Muffin committed:
Refactor: InteractionResponse Subclasses

- Create InteractionResponse subclasses for each Interaction subclass
- Removed unnecessary imports from interaction.Base file
- Edited each method and class description to match the Interaction/-Response subclass
- Altered TypeHints to use the Relevant InteractionResponse subclasses

pre-commit-ci[bot] committed:
style: auto fixes from pre-commit hooks

Gangsta-Muffin committed:
typehint: Fix Pyright errors

Gangsta-Muffin committed:
refactor: added pre-commit ci changed

pre-commit-ci[bot] committed:
style: auto fixes from pre-commit hooks

Gangsta-Muffin committed:
refactor: Removed redundant __slots__

Error was being raised in github

Gangsta-Muffin committed:
Merge branch 'master' of https://github.com/Gangsta-Muffin/interacion_subclasses

teaishealthy commented on nextcord/interactions/application.py:
```diff
@@ -0,0 +1,818 @@
+    def _get_application_options(self, data):
+        options = data
+
+        if len(options) == 0:
```
Empty containers are falsy
```suggestion
        if not options:
```

teaishealthy replied:
Forgot to apply this

teaishealthy commented on nextcord/interactions/message_component.py:
```diff
@@ -447,191 +341,62 @@ async def edit_original_message(
+        if self._original_message is not None:
+            return self._original_message
+
+        # TODO: fix later to not raise?
```
?

Gangsta-Muffin replied:
This is a Todo inserted by someone else - was just copied over. I am guessing it still wants doing at some point

Gangsta-Muffin replied:
--- from old code; Discussed in nextcord support server ---

teaishealthy commented on nextcord/interactions/message_component.py:
```diff
@@ -1,102 +1,44 @@
-    def __repr__(self) -> str:
-        return f"<InteractionAttached {super().__repr__()}>"
 
+class MessageComponentInteraction(Interaction):
```
A lot of this code is duplicated across both classes in application.py. Move into a seperate super class or base.py

teaishealthy commented:
also needs version docs thingies

teaishealthy commented:
🦅

Gangsta-Muffin commented on nextcord/interactions/__init__.py:
```diff
@@ -0,0 +1,15 @@
+
+Interactions collection for the Discord API
+
+:copyright: (c) 2015-2021 Rapptz
```
> this file contains no code from Rapptz

What should I write there instead?

Gangsta-Muffin commented on nextcord/interactions/application.py:
```diff
@@ -0,0 +1,818 @@
+from ..message import Attachment
+from ..webhook.async_ import WebhookMessage, async_context, handle_message_parameters
+from .base import (
+    MISSING,
```
that's what MISSING from .base is. should I do that anyways?

teaishealthy replied:
yes

Gangsta-Muffin commented on nextcord/interactions/application.py:
```diff
@@ -0,0 +1,818 @@
+    ) -> None:
+        self.application_command = app_cmd
+
+    def _get_application_options(self, data):
```
I originally Had `ApplicationAutocompleteInteraction` inherit from `ApplicationCommandInteraction` as the Autocomplete interaction contains similar data as the app command interaction class. In a later commit, I realized I still had some methods I needed to transfer from the base Interaction class, methods that would be compatible for App commands, but not for autocompletes, so I changed the inherit for Autocomplete, and copy pasted the relevant methods Autocomplete was inheriting, and put it in the class.

I suppose a Payload class would be better, which both autocomplete and app command would use? Would reduce redundancy I suppose.

Gangsta-Muffin commented on nextcord/interactions/modal_submit.py:
```diff
@@ -0,0 +1,764 @@
+        return message
+
+
+class ModalSubmitInteractionResponse(InteractionResponse):
```
What code are you referencing specifically?

Gangsta-Muffin committed:
Implement changes made in comments

pre-commit-ci[bot] committed:
style: auto fixes from pre-commit hooks

Gangsta-Muffin commented on nextcord/interactions/application.py:
```diff
@@ -0,0 +1,818 @@
+                suppress_embeds=suppress_embeds,
+            )
+        return await self.followup.send(
+            content=content,  # type: ignore
```
--- from old code; Discussed in nextcord support server ---

Gangsta-Muffin commented on nextcord/interactions/application.py:
```diff
@@ -0,0 +1,818 @@
+        if self._responded:
+            raise InteractionResponded(self._parent)
+
+        defer_type: int = 0
```
--- from old code; Discussed in nextcord support server ---

Gangsta-Muffin commented on nextcord/interactions/application.py:
```diff
@@ -0,0 +1,818 @@
+            The original interaction response message.
+        """
+
+        # TODO: fix later to not raise?
```
--- from old code; Discussed in nextcord support server ---

Gangsta-Muffin committed:
remove false copyright note in __init__

Gangsta-Muffin committed:
pyright and typehint fixes

Gangsta-Muffin committed:
Merge branch 'master' of https://github.com/Gangsta-Muffin/interacion_subclasses

Gangsta-Muffin committed:
pyright and typehint fixes

pre-commit-ci[bot] committed:
style: auto fixes from pre-commit hooks

Gangsta-Muffin committed:
pyright fixes

pre-commit-ci[bot] committed:
style: auto fixes from pre-commit hooks

EmreTech commented on nextcord/client.py:
```diff
@@ -2869,7 +3001,38 @@ class be provided, it should be a subclass of :py:class:`Interaction`.
+
+        Warns
+        --------
+        ~FutureWarning
```
I don't think it's worth documenting this warning, considering that other functions throughout the library do not either.

EmreTech commented on nextcord/client.py:
```diff
@@ -2869,7 +3001,38 @@ class be provided, it should be a subclass of :py:class:`Interaction`.
+            in the event that defining this parameter was a mistake, and other parameters were to be defined instead.
+        """
+        # check if they are creating custom interaction subclasses
+        if cls != Interaction:
```
```suggestion
        if cls is not Interaction:
```
`!=` is usually used with literals.

EmreTech commented on nextcord/client.py:
```diff
@@ -2869,7 +3001,38 @@ class be provided, it should be a subclass of :py:class:`Interaction`.
+        # check if they are creating custom interaction subclasses
+        if cls != Interaction:
+            warnings.warn(
+                "The 'cls' parameter was given a custom Interaction object. "
```
```suggestion
                "The 'cls' parameter was given a custom Interaction class. "
```

EmreTech commented on nextcord/components.py:
```diff
@@ -77,6 +77,10 @@ def __repr__(self) -> str:
     def _raw_construct(cls, **kwargs) -> Self:
         self = cls.__new__(cls)
         for slot in get_slots(cls):
+            if slot == "type" and getattr(self, slot, None) is not None:
+                # nextcord/issues/1103
+                continue
```
Wasn't this fixed by #1104?

Gangsta-Muffin replied:
Possibly? I haven't been keeping track on more recent PRs. Please let me know what I should do with this code, provided changes are in order :)

EmreTech replied:
All that PR did when it was merged was remove `"type"` from the `__slots__` of the `Component` class. As long as your fork is up to date with the upstream and you remove this code I commented on, all should be good.

Gangsta-Muffin committed:
Update nextcord/application_command.py

Co-authored-by: Emre Terzioglu <50607143+EmreTech@users.noreply.github.com>

Gangsta-Muffin committed:
Update nextcord/client.py

Co-authored-by: Emre Terzioglu <50607143+EmreTech@users.noreply.github.com>

Gangsta-Muffin committed:
Update nextcord/application_command.py

Co-authored-by: Emre Terzioglu <50607143+EmreTech@users.noreply.github.com>

Gangsta-Muffin committed:
Update nextcord/application_command.py

Co-authored-by: Emre Terzioglu <50607143+EmreTech@users.noreply.github.com>

Gangsta-Muffin committed:
Update nextcord/application_command.py

Co-authored-by: Emre Terzioglu <50607143+EmreTech@users.noreply.github.com>

Gangsta-Muffin committed:
Update nextcord/application_command.py

Co-authored-by: Emre Terzioglu <50607143+EmreTech@users.noreply.github.com>

Gangsta-Muffin commented on nextcord/types/interactions.py:
```diff
@@ -202,16 +202,46 @@ class Interaction(TypedDict):
     token: str
     version: int
-    data: NotRequired[InteractionData]
+    # data: NotRequired[InteractionData]
```
Because the subclasses recieved their own attribute for data. I did this partially due to typehinting issues, but also because the PingInteraction class (not currently in nextcord, but exists in the dev docs) does not have a data key. (But I mainly did it due to typehint issues - since the `InteractionData` type is a Union of all InteractionData types, and in the various subclasses, pyright would complain in various places, that various code was okay provided if the typehint was the correct one, but would have issues if the type was that used for the other subclasses)

EmreTech replied:
Yeah, I forgot about that while reviewing this PR.

Gangsta-Muffin committed:
Update nextcord/interactions/application.py

Co-authored-by: Emre Terzioglu <50607143+EmreTech@users.noreply.github.com>

Gangsta-Muffin committed:
Update nextcord/interactions/application.py

Co-authored-by: Emre Terzioglu <50607143+EmreTech@users.noreply.github.com>

Gangsta-Muffin committed:
Update nextcord/interactions/application.py

Co-authored-by: Emre Terzioglu <50607143+EmreTech@users.noreply.github.com>

Gangsta-Muffin committed:
Update nextcord/interactions/application.py

Co-authored-by: Emre Terzioglu <50607143+EmreTech@users.noreply.github.com>

Gangsta-Muffin committed:
Update nextcord/client.py

Co-authored-by: Emre Terzioglu <50607143+EmreTech@users.noreply.github.com>

Gangsta-Muffin commented on nextcord/interactions/application.py:
```diff
@@ -0,0 +1,808 @@
+    def _from_data(self, data: ApplicationCommandPayload) -> None:
+        super()._from_data(data=data)
+
+        self.data: InteractionData = data.get("data")
```
They subclasses do in fact use the correct types. The imports follow similar patter to the following: 
`from ..types.interactions import ApplicationCommandInteractionData as InteractionData`

EmreTech replied:
Ah, I missed that part. I don't see why it should be aliased to `InteractionData`, it makes the code a little confusing to read through.

Gangsta-Muffin replied:
I aliased them for the following reasons:
- It was aliased in the original Interaction file
- For the sake of reducing the length of the typehints since they are quite big
- For the sake of simplifying things a bit

EmreTech replied:
1. No it wasn't? I just looked at the source code, and only the `Interaction` typed dict was aliased to not conflict with the actual `Interaction` class defined in the file.
2. That's not a strong argument to use this alias. To me, the whole point of type hints is not only for static type checkers, but also for people reading the code to understand what a specific variable/attribute/parameter is. Shortening it like how you did it here makes it less obvious what specifically the type of the `data` attribute is. Plus, we already have linters to take care of long lines.
3. How would this alias simplify things? I don't understand.

Gangsta-Muffin replied:
Regarding point 3: My thought was mainly to Simply the class name, since `ApplicationCommandInteractionData` was very long and a bit overwhelming, and I thought the alias simplified things and in result, allowed for better understanding of the code - Albeit, based on your counter arguments, I can see the flaws and reasons for why this isn't a good idea; will be changing the code

Gangsta-Muffin commented on nextcord/interactions/application.py:
```diff
@@ -0,0 +1,808 @@
+            self._get_application_options(options) if options else []
+        )
+
+    def _set_application_command(
+        self, app_cmd: Union[SlashApplicationSubcommand, BaseApplicationCommand]
+    ) -> None:
+        self.application_command = app_cmd
```
This is part of old code not written by me. From searching through the library, by the looks of things, it gets used in `CallbackMixing.invoke_callback_with_hooks()`. Im guessing it serves a purpose, but since its not my code, I've not looked into it

Gangsta-Muffin commented on nextcord/application_command.py:
```diff
@@ -1236,6 +1248,46 @@ def error_name(self) -> str:
+
+    __slots__: Tuple[str, ...] = ("data", "value", "type", "name", "focused")
+
+    def __init__(self, data) -> None:
```
This type doesn't exist - Are you suggesting I should make a new type for this?

EmreTech replied:
https://github.com/nextcord/nextcord/blob/5d2eff1810edd76359b66f37f731537a520c3d2c/nextcord/types/interactions.py#L128-L135

It does exist, and it's already imported in this module.

Gangsta-Muffin replied:
Must have missed that somehow. Either way, as part of that list, is the class  `_ApplicationCommandInteractionDataOptionSubcommand`, which the SlashOptionData class doesn't use. (It may possibly not use Snowflake either but I can't say this for sure). Would it be worth making a new dedicated type for this? or is there some typehinting approach I can take, to reassure pyright, that this class does, indeed, not get used?

EmreTech replied:
Well, it does get used to tell Pyright what keys and types of values a specific dictionary has. You could make a new union without the subcommand one, but I'm not sure what to call it.

Gangsta-Muffin committed:
Type fix for Application Commands

teaishealthy commented on nextcord/interactions/application.py:
```diff
@@ -0,0 +1,808 @@
+        self.app_command_id: Snowflake = self.data["id"]
+
+        options = self.data.get("options")
+        self.options: List[SlashOptionData] = (
+            self._get_application_options(options) if options else []
+        )
```
preview is broken for some reason, but what Emre said.
```suggestion
        options = self.data.get("options", [])
        self.options: List[SlashOptionData] = (
            self._get_application_options(options)
        )
```

teaishealthy commented on nextcord/interactions/application.py:
```diff
@@ -0,0 +1,808 @@
+            SlashOptionData,
+        )
+
+        return [SlashOptionData(option) for option in options]
```
pyright is not happy with this because you are changing the type of an already assigned variable.

teaishealthy commented on nextcord/interactions/application.py:
```diff
@@ -0,0 +1,808 @@
+        self._parent._state.store_modal(modal, self._parent.user.id)  # type: ignore
+
+
+class ApplicationAutocompleteInteraction(Interaction):
```
@Gangsta-Muffin ?

ooliver1 commented on nextcord/client.py:
```diff
@@ -2869,7 +3001,38 @@ class be provided, it should be a subclass of :py:class:`Interaction`.
+            in the event that defining this parameter was a mistake, and other parameters were to be defined instead.
+        """
+        # check if they are creating custom interaction subclasses
+        if cls is not Interaction:
```
This warning seems a bit weird, the signature for `get_interaction` seems to imply you should pass a custom interaction when passing other custom interaction classes. The better check would be to look at the other classes and if none are custom, but that just seems a bit excessive IMO.
```python
class Client(nextcord.Client):
    def get_interaction(self, data: InteractionPayload):
        return super().get_interaction(
            data,
            application_autocomplete_interaction=CustomAutocompleteInteraction,
            application_interaction=CustomAppCmdInteraction,
            message_component_interaction=CustomComponentInteraction,
            modal_submit_interaction=CustomModalSubmitInteraction,
            cls=CustomInteraction,
        )
```

ooliver1 commented on nextcord/ui/item.py:
```diff
@@ -2,9 +2,11 @@
-from ..interactions import ClientT, Interaction
+from ..interactions.base import ClientT, Interaction
+from ..interactions.message_component import MessageComponentInteraction
+from ..interactions.modal_submit import ModalSubmitInteraction
```
`ItemCallbackType` should probably use the component interaction class not the base one:
```python
Argument of type "(self: Self@View, button: Button[Unknown], interaction: CustomComponentInteraction) -> Coroutine[Any, Any, None]" cannot be assigned to parameter of type "ItemCallbackType[Button[V@button], ClientT@button]"
  Type "(self: Self@View, button: Button[Unknown], interaction: CustomComponentInteraction) -> Coroutine[Any, Any, None]" cannot be assigned to type "ItemCallbackType[Button[V@button], ClientT@button]"
    Parameter 3: type "Interaction[ClientT@button]" cannot be assigned to type "CustomComponentInteraction"
      "Interaction[ClientT@button]" is incompatible with "CustomComponentInteraction"Pylance[reportGeneralTypeIssues](https://github.com/microsoft/pyright/blob/main/docs/configuration.md#reportGeneralTypeIssues)
```

ooliver1 reviewed (changes requested):
Tested. Only issues were the warning and button type error mentioned in these comments. https://paste.nextcord.dev/?id=1702153210128711
